File: apps/dotnet/src/Cli/Server/ServerWire.cs
// Copyright 2026 Shazron Abdullah and Bunyi contributors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System.Text;
using System.Text.Json;
using Bunyi.Cli.Protocol;

namespace Bunyi.Cli.Server;

internal sealed record WireRequest(int ProtocolVersion, string Action, CommandRequest? Request = null, string? JobId = null, bool Detached = false);

internal static class ServerWire
{
    internal const int Version = 1;
    internal static readonly JsonSerializerOptions Json = new(JsonSerializerDefaults.Web);
    internal static async Task WriteAsync(Stream stream, object value, CancellationToken ct)
    {
        var bytes = JsonSerializer.SerializeToUtf8Bytes(value, Json);
        await stream.WriteAsync(bytes, ct);
        await stream.WriteAsync(new byte[] { 10 }, ct);
        await stream.FlushAsync(ct);
    }

    // Bounded framing prevents a peer from allocating an unbounded ReadLine string.
    internal static async Task<string?> ReadAsync(Stream stream, CancellationToken ct)
    {
        using var line = new MemoryStream();
        var single = new byte[1];
        while (await stream.ReadAsync(single, ct) != 0)
        {
            if (single[0] == 10) return Encoding.UTF8.GetString(line.GetBuffer(), 0, (int)line.Length);
            if (line.Length >= 1024 * 1024) throw new ServerException("invalid_arguments", "The server request exceeds the one-megabyte limit.");
            line.WriteByte(single[0]);
        }
        return line.Length == 0 ? null : throw new IOException("Incomplete server message.");
    }

    internal static Dictionary<string, object?> Result(string operation, string? id = null, string type = "result") => new()
    {
        ["schemaVersion"] = 1, ["ok"] = true, ["type"] = type,
        ["operation"] = operation, ["operationId"] = id ?? Guid.NewGuid().ToString("N")
    };

    internal static Dictionary<string, object?> Error(string operation, string? id, string code, string message)
    {
        var value = Result(operation, id, "error");
        value["ok"] = false;
        value["error"] = new { code, message };
        value["exitCode"] = code == "invalid_arguments" ? 2 : 4;
        return value;
    }

    internal static string? Text(Dictionary<string, object?> value, string key) => value.TryGetValue(key, out var item) ? item?.ToString() : null;
}

File: apps/macos/CLI/ServerWire.swift
// Copyright 2026 Shazron Abdullah and Bunyi contributors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import Darwin
import Foundation

struct ServerError: LocalizedError, Sendable {
    let code: String
    let message: String

    var errorDescription: String? { message }
}

enum ServerWire {
    static let version = 1
    static let maximumMessageBytes = 1 << 20

    static func encode(_ value: CLIMessage) throws -> Data {
        guard JSONSerialization.isValidJSONObject(value) else {
            throw ServerError(
                code: "server_protocol_mismatch",
                message: "Bunyi tried to send an invalid server message.")
        }
        var data = try JSONSerialization.data(
            withJSONObject: value, options: [.sortedKeys])
        data.append(0x0A)
        return data
    }

    static func decode(_ data: Data) throws -> CLIMessage {
        guard let value = try JSONSerialization.jsonObject(with: data)
                as? CLIMessage else {
            throw ServerError(
                code: "server_protocol_mismatch",
                message: "The server message is not a JSON object.")
        }
        return value
    }

    static func write(_ data: Data, to descriptor: Int32) throws {
        try data.withUnsafeBytes { raw in
            guard let start = raw.baseAddress else { return }
            var written = 0
            while written < raw.count {
                let count = Darwin.write(
                    descriptor, start.advanced(by: written), raw.count - written)
                if count > 0 {
                    written += count
                } else if count < 0, errno == EINTR {
                    continue
                } else {
                    throw POSIXError(POSIXErrorCode(rawValue: errno) ?? .EIO)
                }
            }
        }
    }

    /// Reads one bounded newline-framed message. A negative timeout waits
    /// indefinitely; a timeout is reported as ETIMEDOUT.
    static func read(from descriptor: Int32, timeoutMilliseconds: Int32 = -1)
        throws -> Data? {
        var result = Data()
        var byte: UInt8 = 0
        while true {
            if timeoutMilliseconds >= 0 {
                var item = pollfd(
                    fd: descriptor, events: Int16(POLLIN), revents: 0)
                let ready = Darwin.poll(&item, 1, timeoutMilliseconds)
                if ready == 0 {
                    throw POSIXError(.ETIMEDOUT)
                }
                if ready < 0 {
                    if errno == EINTR { continue }
                    throw POSIXError(POSIXErrorCode(rawValue: errno) ?? .EIO)
                }
            }
            let count = Darwin.read(descriptor, &byte, 1)
            if count == 0 {
                if result.isEmpty { return nil }
                throw POSIXError(.ECONNRESET)
            }
            if count < 0 {
                if errno == EINTR { continue }
                throw POSIXError(POSIXErrorCode(rawValue: errno) ?? .EIO)
            }
            if byte == 0x0A { return result }
            guard result.count < maximumMessageBytes else {
                throw ServerError(
                    code: "invalid_arguments",
                    message: "The server request exceeds the one-megabyte limit.")
            }
            result.append(byte)
        }
    }

    static func request(_ request: CLIRequest, action: String,
                        detached: Bool = false) -> CLIMessage {
        var arguments: CLIMessage = [:]
        for (key, value) in request.arguments {
            arguments[key] = value ?? NSNull()
        }
        return [
            "protocolVersion": version,
            "action": action,
            "detached": detached,
            "request": [
                "operation": request.operation,
                "operationId": request.operationID,
                "arguments": arguments,
            ] as CLIMessage,
        ]
    }

    static func parseRequest(_ value: CLIMessage) throws -> CLIRequest {
        guard let request = value["request"] as? CLIMessage,
              let operation = request["operation"] as? String,
              let operationID = request["operationId"] as? String,
              !operation.isEmpty, !operationID.isEmpty,
              operationID.count <= 128,
              let rawArguments = request["arguments"] as? CLIMessage else {
            throw ServerError(
                code: "invalid_arguments",
                message: "The server command is missing required fields.")
        }
        var arguments: [String: String?] = [:]
        for (key, value) in rawArguments {
            if value is NSNull {
                arguments[key] = .some(nil)
            } else if let value = value as? String {
                arguments[key] = value
            } else {
                throw ServerError(
                    code: "invalid_arguments",
                    message: "The server command contains an invalid argument.")
            }
        }
        return CLIRequest(
            operation: operation, arguments: arguments,
            operationID: operationID)
    }

    static func hello() -> CLIMessage {
        ["protocolVersion": version]
    }
}
